crates/core/src/instance.rs

```diff
@@ -267,6 +273,60 @@ impl Instance {
         Ok(())
     }
 
+    fn start_holding_loop(&mut self, context: Arc<Context>) {
+        let (kill_sender, kill_receiver) = crossbeam_channel::unbounded();
+        self.kill_switch_holding = Some(kill_sender);
+        thread::Builder::new()
+            .name(format!(
+                "holding_loop/{}",
+                ProcessUniqueId::new().to_string()
+            ))
+            .spawn(move || {
+                while kill_receiver.try_recv().is_err() {
+                    log_trace!(context, "Checking holding queue...");
+                    loop {
+                        let maybe_holding_workflow = context
+                            .state()
+                            .expect("Couldn't get state in run_pending_validations")
+                            .dht()
+                            .next_queued_holding_workflow();
+                        if let Some(pending) = maybe_holding_workflow {
+                            log_debug!(context, "Found queued validation: {:?}", pending);
+                            context.block_on(pop_next_holding_workflow(
+                                pending.clone(),
+                                context.clone(),
+                            ));
```

Comment:
Can't we just pop in one step, rather than checking the next queued item and then popping? Having two steps creates the possibility of a very slim race condition.

Reply:
It is hard to make pop an atomic action like you would if the queue would be owned by the local thread, but the queue lives in the instance's state and can only get mutated by reducers run in the redux action thread. The race condition is taken care of by having the item you think you're popping be an argument of the pop action. If the head is not that item anymore the reducer will not pop the next one.